pkg/networkservice/connectioncontext/dnscontext/client.go

```diff
@@ -0,0 +1,126 @@
+	"os"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/dnscontext"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+)
+
+type dnsContextClient struct {
+	cancelMonitoring context.CancelFunc
+	chainContext     context.Context
+	monitorContext   context.Context
+	coreFilePath     string
+	getCallOptions   func() []grpc.CallOption
+	dnsConfigManager dnscontext.Manager
+	monitorClient    networkservice.MonitorConnectionClient
+	executor         serialize.Executor
+}
+
+// NewClient creates a new DNS client chain component. Setups all DNS traffic to the localhost. Monitors DNS configs from connections.
+func NewClient(chainContext context.Context, coreFilePath, resolveConfigPath string, monitorClient networkservice.MonitorConnectionClient, getMonitorCallOptions func() []grpc.CallOption) networkservice.NetworkServiceClient {
+	c := &dnsContextClient{
+		coreFilePath:     coreFilePath,
+		chainContext:     chainContext,
+		dnsConfigManager: dnscontext.NewManager(),
+		monitorClient:    monitorClient,
+		getCallOptions:   getMonitorCallOptions,
+	}
+	if r, err := dnscontext.OpenResolveConfig(resolveConfigPath); err != nil {
+		logrus.Errorf("DnsContextClient: can not load resolve config file. Path: %v. Error: %v", resolveConfigPath, err.Error())
+	} else {
+		c.dnsConfigManager.Store("", &networkservice.DNSConfig{
+			SearchDomains: r.Value(dnscontext.AnyDomain),
+			DnsServerIps:  r.Value(dnscontext.NameserverProperty),
+		})
+		r.SetValue(dnscontext.NameserverProperty, "127.0.0.1")
+		if err := r.Save(); err != nil {
+			logrus.Errorf("DnsContextClient: can not update resolve config file. Error: %v", err.Error())
+		}
+	}
+	return c
+}
+
+func (c *dnsContextClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	rv, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
+	}
+	c.monitorContext, c.cancelMonitoring = context.WithCancel(c.chainContext)
+	c.executor.AsyncExec(c.monitorConfigs)
+	return rv, err
+}
+
+func (c *dnsContextClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	r, err := next.Client(ctx).Close(ctx, conn, opts...)
+	if err != nil {
+		return nil, err
+	}
+	c.cancelMonitoring()
+	return r, err
+}
+
+func (c *dnsContextClient) monitorConfigs() {
+	steam, err := c.monitorClient.MonitorConnections(c.monitorContext, &networkservice.MonitorScopeSelector{}, c.getCallOptions()...)
+	if err != nil {
+		c.executor.AsyncExec(c.monitorConfigs)
+		return
+	}
+	for {
+		if c.monitorContext.Err() != nil {
+			return
+		}
+		event, err := steam.Recv()
+		if err != nil {
+			c.executor.AsyncExec(c.monitorConfigs)
```

Comment:
I think at least logging of err are required.

Reply:
Thanks, fixed.